### PR TITLE
Add a DCPower LabVIEW MeasurementLink example

### PR DESCRIPTION
**Pull request by pbirkhol-ni**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Adds a DCPower MeasurementLink example showing best practice for LabVIEW measurements.

Note that I created this example using the MeasurementLink project wizard. I borrowed as much as I could from the DCPower python example (the LabVIEW UI, readme, code flow, etc).

In the comments below, I will add a non comprehensive set of screenshots and also comment on code that I'd like additional input. For a full review, you should really pull down the example and open it in LabVIEW.

### Why should this Pull Request be merged?

Customers will need at least one full featured example to be successful with MeasurementLink.

### What testing has been done?

I installed MeasurementLink 23.0, InstrumentStudio 23.0, LabVIEW 2020SP1, TestStand 2021, and DCPower 23.0 on a clean machine. I installed the VI Packages for MeasurementLink. Then I copied over the example and tested it in both InstrumentStudio and LabVIEW.

### CODE REVIEW

**Comment by pbirkhol-ni** [comment]
Type specializations for the pin control:
image

**Comment by pbirkhol-ni** [comment]
Measurement details:
image

**Comment by pbirkhol-ni** [comment]
Measurement Logic:
image

**Comment by pbirkhol-ni** [comment]
Create and Register Sessions code module:
image

**Comment by pbirkhol-ni** [comment]
Unregister and Destroy Sessions code module:
image

**Comment by pbirkhol-ni** [comment]
Update Pin Map code module
image

**Comment by pbirkhol-ni** [comment]
DC Power specific logic. This gets called from the `Measurement Logic.vi`.
image

**Comment by pbirkhol-ni** [comment]
Areas of note:
- When creating sessions in the code module, I create them and then immediately detach from them. I think that is the right behavior, but it is hard to map directly to the python example since python auto handles this.
image

- When reserving and unreserving sessions, I typically wire the Response Data of reserve to the Request Data of unreserve. This cleans up the diagram considerably, but is probably bad practice. Note that there is a coercion dot that shows the type mismatch.
image

- When creating sessions in the code module, I don't create a session if it already exists and I only register the sessions that I created. So for example, if the code module is supposed to create two sessions but one already exists, I will skip it and only create and register the session that doesn't exist. This is different from how python behaves (python assumes that none of the sessions exist). I don't know that it matters either way, but the LabVIEW code seems safer.
image

- In the TestStand sequence, I don't just assign the PinMapPath file global to the PinMapId temporary global like the python examples. Instead, I read the pin map id from the code module, assign that to a PinMapId file global, and then assign that to the PinMapId temporary global. According to Jonathan Mendez, this should be considered best practice and I filed a bug to python to update the examples.
- The code modules are part of the `NI-DCPower Source DC Voltage.lvlib`. This seems okay, but perhaps could cause build problems if/when trying to build the code modules into a PPL/exe?

**Comment by dixonjoel** [comment]
Why do the calls into resolve service have two copies of the same service class?

**Comment by dixonjoel** [comment]
I get this type conflict when opening in TestStand. I suspect it's a problem with my dev machine - I didn't check which MeasurementLink I even have installed, but just wanted to verify you've been able to open the sequence on a test machine and ask which version(s) of MeasurementLink work with this sequence file.

image

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Source DC Voltage.lvproj
Comment (dixonjoel) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Source DC Voltage.lvproj:1]: Why the difference in file name for this .lvproj file? All the others are NIDCPowerSourceDCVoltage*.

**Comment by dixonjoel** [comment]
How much failure testing have you done to see how well the example handles and communicates error states? E.g. what if the discovery service is not running when you run the example? What if the pin map service is not installed? etc.

I don't think this kind of testing needs to hold up the submission of the example, but I think we should try a few of the common errors a user may experience and see how the error reporting is and if it's reasonable to help the user understand what's wrong.

**Comment by dixonjoel** [comment]
Have you tested the example by building it into a .exe to flush out any errors that may occur in the runtime but not in the dev environment?

File: Source/Example Measurements/README.md
Comment (dixonjoel) [inline Source/Example Measurements/README.md:18]: I think this should be 'NI MAX'.

**Comment by dixonjoel** [comment]
Did you intentionally leave out the .measui and .measproj from the Python example? I don't think it has to be included in this initial checkin, but I think it would be nice to show that a LabVIEW service can use .measui or LV UI just like a Python service.

**Comment by bkeryan** [comment]
> Why do the calls into resolve service have two copies of the same service class?
> 
> Edit: It looks like it's the provided interface and service class. Are these always the same for a LV service?

The provided interface and service class are the same for "singleton" services like Session Management.

**Comment by pbirkhol-ni** [comment]
> I get this type conflict when opening in TestStand. I suspect it's a problem with my dev machine - I didn't check which MeasurementLink I even have installed, but just wanted to verify you've been able to open the sequence on a test machine and ask which version(s) of MeasurementLink work with this sequence file.
> 
> image
> 
> Followed by image

I wrote the example targeting the new NI_MeasurementLink.ini file. Once that is in the installers you shouldn't get that message. However, we may start getting that message with the python examples since they target the old name.

**Comment by pbirkhol-ni** [comment]
> Did you intentionally leave out the .measui and .measproj from the Python example? I don't think it has to be included in this initial checkin, but I think it would be nice to show that a LabVIEW service can use .measui or LV UI just like a Python service.

It wasn't intentional. I would be surprised if someone writing a LabVIEW measurement uses a .measui UI, but it is definitely possible. I can add one.

**Comment by pbirkhol-ni** [comment]
> Have you tested the example by building it into a .exe to flush out any errors that may occur in the runtime but not in the dev environment?

No, I haven't. I will test that. It gets a little tricky with the code modules. However, I can test that the measurement works. The strategy for deploying the code modules will be decided later.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Source DC Voltage.lvproj
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Source DC Voltage.lvproj:1]: I created the example using the measurement template before I started stealing files from the python example. I will go ahead and change the project and library name to be consistent.

**Comment by pbirkhol-ni** [comment]
> > Why do the calls into resolve service have two copies of the same service class?
> > Edit: It looks like it's the provided interface and service class. Are these always the same for a LV service?
> 
> The provided interface and service class are the same for "singleton" services like Session Management.

I followed the python example and provided the interface and service class. Technically it should work with one or the other, but I think it is safer to provide both. Since they are the same string, I could have a single string and bundle it twice into the cluster, but I kind of like the cluster constant better.

**Comment by bkeryan** [comment]
Do not pass timeout=0 to ReserveSessions and ReserveAllRegisteredSessions. Either choose a timeout or pass timeout=-1 for infinite. The Python examples pass timeout=-1.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:23]: Do not pass timeout=0 to ReserveSessions and ReserveAllRegisteredSessions. Either choose a timeout or pass timeout=-1 for infinite. The Python examples pass timeout=-1.

This also applies to Measurement Logic.vi.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: Each call to the measurement opens three gRPC channels: discovery, session management, and grpc-device. If we don't already have a LV gRPC channel pool in the backlog, we should add it.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:17]: Add "insecure" to the VI name to indicate that it uses the insecure port

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:19]: Add some form of NI-DCPower to the VI name and update it to use `niDCPower gRPC Device Server.ctl`

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: FYI: Consider doing the ResolveService calls in parallel, at least for Measurement Logic.vi.

**Comment by pbirkhol-ni** [comment]
> > Have you tested the example by building it into a .exe to flush out any errors that may occur in the runtime but not in the dev environment?
> 
> No, I haven't. I will test that. It gets a little tricky with the code modules. However, I can test that the measurement works. The strategy for deploying the code modules will be decided later.

I built it into an exe, put it in the %Program Data% directory, and everything just worked.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:22]: The behavior of the Python TestStand fixture is to ignore "session_exists" and always initialize. That way, if the session exists, the grpc-device server will return an error.

**Review by bkeryan — CHANGES_REQUESTED** [review]

**Comment by pbirkhol-ni** [comment]
> > Did you intentionally leave out the .measui and .measproj from the Python example? I don't think it has to be included in this initial checkin, but I think it would be nice to show that a LabVIEW service can use .measui or LV UI just like a Python service.
> 
> It wasn't intentional. I would be surprised if someone writing a LabVIEW measurement uses a .measui UI, but it is definitely possible. I can add one.

I added a .measui. Unfortunately, LabVIEW only looks for UI files in a sub directory so the .measui file and .measproj files are in different directories. I don't think this is a problem, but I can file an issue so that LabVIEW looks in more places for the UI files.

**Comment by pbirkhol-ni** [comment]
> How much failure testing have you done to see how well the example handles and communicates error states? E.g. what if the discovery service is not running when you run the example? What if the pin map service is not installed? etc.
> 
> I don't think this kind of testing needs to hold up the submission of the example, but I think we should try a few of the common errors a user may experience and see how the error reporting is and if it's reasonable to help the user understand what's wrong.

If you specify the wrong resource name, you get a very descriptive error. If the session manager or pin map service aren't installed the error is less clear but it would be possible to figure it out. Overall, all errors get reported, but they may or may not be helpful. I don't think they are any better/worse than errors for other examples.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:23]: I talked to Jonathan Mendez about this. He didn't make a recommendation, but told me that a 0 timeout means that we will try to reserve a session and if it is already reserved it will immediately throw an error. That seems reasonable to me, especially in the code module where we expect the sessions to be available. I would argue that -1 is the worst behavior because it will cause your application to hang and you won't know why. If my understanding is incorrect, let me know.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:22]: Okay. I will make that change.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: I will check and add one if not. I agree that this isn't ideal.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:23]: Ok, how about 60000 (1 min) for the measurement and 0 for the code module? We should do the same in Python.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:19]: Interesting. I didn't know that ctrl existed. Why don't the DCPower MeasurmentLink VIs use it? 

Do you still think this conversion VI should use the typedef ctrl even though the Init and Attach VIs don't use it? It will result in the a coercion dot, which isn't the end of the world.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:23]: Sounds reasonable to me. I while file an issue to python.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (dixonjoel) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:19]: They should use it.
[AB#2287962](https://dev.azure.com/ni/DevCentral/_workitems/edit/2287962): NI-DCPower gRPC APIs shall use typedef control `niDCPower gRPC Device Server.ctl` in LabVIEW API for initialize and attach.

Comment (dixonjoel) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:19]: I don't know if it can be changed now since NI-DCPower has shipped, but I don't see a reason for us not to use it.

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: I created a new feature in the Feature Debt epic.

Comment (pbirkhol-ni) [reply Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: I parallelized it in the measurement logic.

**Comment by bkeryan** [comment]
Let's also add comments explaining the timeouts, if you didn't already (I didn't open the VIs). Something like this:

- teststand_fixture.py: This code module sets up the sessions, so error immediately if they are in use.
- measurement.py: If another measurement is using the session, wait for it to complete. Specify a timeout to aid in debugging missed unreserve calls. Long measurements may require a longer timeout.

**Comment by pbirkhol-ni** [comment]
@bkeryan @dixonjoel I have been discussing this example with Ryan and Jack the last few days. After looking into a few additional options, we decided to leave it as is. I have responded to all of the PR comments and am now just waiting on approval.

**Review by dixonjoel — APPROVED** [review]

File: Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib
Comment (bkeryan) [inline Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/NIDCPowerSourceDCVoltage.lvlib:30]: Cosmetic: the cluster input to DiscoveryService ResolveService.vi appears to be running right-to-left under the subVI. 
image

**Review by bkeryan — APPROVED** [review]

**Review by aepete — APPROVED** [review]

**Pull request merged**
